### PR TITLE
parse nan inf and -inf in json

### LINKED ISSUE #10700: Treatment of Inf/-Inf/NaN in the json module
The JSON specification doesn't include support for the special float values (Inf, -Inf, and NaN), so their treatment in different languages and libraries vary greatly. This is the current behavior for the `json` module:

- `Inf` produces `Infinity` for JS backend, otherwise `inf`
- `-Inf` produces `-Infinity` for JS backend, otherwise `inf`
- `NaN` produces `NaN` for JS backend, otherwise `nan`

It should be noted that none of these are accepted by `json.parseJson`, so the `json` module produces `json` that it can't parse itself.

Since interoperability with other languages is important for JSON, here's a quick rundown of the default behavior of some other languages:

- JavaScript: produces `null` for all, only parses strictly valid JSON
- Python: Produces `Infinity`, `-Infinity` and `NaN` and parses them as well (case sensitive)
- Newtonsoft (C#): Produces `"Infinity"`, `"-Infinity"`, `"NaN"` (strings)
- Ruby: Throws an exception for both parsing and generating

IMO the default behavior should be spec compliance, which in this case means throwing an exception. An option to produce the symbols `Infinity`/`-Infinity`/`NaN` might be a good idea, but then it should be allowed during parsing as well.

### PR DESCRIPTION
**Pull request by krux02**
fixes #10700 

It's clean. Ready for review.

### CODE REVIEW

**Comment by krux02** [comment]
@dom96 The json module also supports integers. I am pretty sure that is out of official json spec as well.

File: lib/pure/json.nim [hunk 1]
```diff
@@ -656,6 +656,8 @@ proc escapeJson*(s: string): string =
   result = newStringOfCap(s.len + s.len shr 3)
   escapeJson(s, result)
 
+proc toUgly*(result: var string, node: JsonNode)
```
Comment (Araq) [inline lib/pure/json.nim]: This breaks Nim's effect inference.

File: lib/pure/parsejson.nim [hunk 2]
```diff
@@ -324,35 +324,54 @@ proc skip(my: var JsonParser) =
       break
   my.bufpos = pos
 
+proc substrEq(s: cstring, pos: int, substr: string): bool =
+  ## copy of strutils.substrEq, because that function isn't exported.
```
Comment (Araq) [inline lib/pure/parsejson.nim]: Move it to ``cstrutils.nim`` then?

File: lib/pure/json.nim (hunk 1 above)
Comment (krux02) [reply lib/pure/json.nim]: I can reorder the procedures, no problem, but that makes the diff larger.

**Comment by krux02** [comment]
feedback has been applied.

**Comment by recloser** [comment]
If you really want to follow what Python does and make the parser/dumper standards non-compliant by default (which I think is a pretty bad idea), then I think you should at least make it an option to parse in a strictly standards compliant manner, similarly to what @GULPF suggested in the related issue.

**Comment by krux02** [comment]
@recloser I am sorry, but I cannot come up with any realistic situation where silently interpreting ``Infinity`` as a floating point infinity value can be the cause of any problem. On the other hand writing ``Infinity`` when on the other side is a client that does not understand ``Infinity`` might be a problem. But that is something that happened before this patch already and this patch does not try to fix that issue. It just makes the printing a bit more consistent (same result on all backends).

**Comment by andreaferretti** [comment]
It can be a cause of problems when processing some data files, not realizing early that they contain errors (because the parser does not fail) and then realizing this when it is too late (maybe some data has been corrupted)

**Comment by krux02** [comment]
@andreaferretti what errors are you talking about?

**Comment by andreaferretti** [comment]
Well, I am answering to

> I cannot come up with any realistic situation where silently interpreting `Infinity` as a floating point infinity value can be the cause of any problem

Imagine you are building some data processing system (some step of an ETL). For some reason, you get corrrupted data files, where a previous system had generated a JSON containing `Infinity`. If you are strict, you will get an exception, and then you have a possibility to figure out what's wrong with the input files, and possibly correct them at the source. If you are lenient, you may process invalid data and then go on with the processing, with unpredictable consequences (corrupt data? initiate a payment? send the missile?).

Of course, this is all very hypothetical, I was just pointing that there is value in being strict, and that it is not hard to concoct a scenario where silently interpreting `Infinity` as a floating point value can cause bugs.

**Comment by Araq** [comment]
I agree with @recloser, we should put this behind a flag.

**Comment by recloser** [comment]
Wrt to parsing I pretty much agree with what @andreaferretti said. I'd rather deal with an exception and fix whatever's generating the invalid data if possible, or kindly inform whoever's sending me that whatever they sent me isn't actually a valid JSON document. The location info from the exception can help in both cases.

As for printing, it'd be better to simply make both backends behave correctly, by printing either JS style `null` or quoted strings. This is a standardized data interchange format outputting an a version unparseable by the vast majority of libs implementing it seems to be missing the point of that pretty hard to me. I don't think anybody actually wants this custom flavor of JSON to end up in their db or data files.

**Comment by krux02** [comment]
Imagine a system that creates for some reason but intentionally the floating point infinity value. This result is then stored in some result object that need to be sent to over the network via json and at the other end the object needs to be deserialized again reconstructing the infinity value. Now there is either a parsing error, because ``null`` isn't a floating point value, or ``null`` is mapped incorrectly to ``0``, ``NaN`` or ``-inf``. Either way the value did not survive the transmission on the network. Some hours of debugging to find out where the ``NaN`` comes from  you find out that the json serializer serializes ``Inf`` as ``null`` and then it will be deserialized as ``Nan``, just because you forgot some flag during serialization.

No I won't support any ideas that break by default some states of a floating point number. If you want a flag to be 100% spec compliant, I am ok with that, but not by default.  ``Infinity`` is not corrupt data. ``NaN`` is not corrupted data either, but serializing any floating point value as ``null`` is corruped data.

I also don't like the idea of a program crash, just bacause the serializer refuses to serialize certain states of a number.

Btw, we are already not spec compiant. To enable spec compliance the parsing of int64 needs to be disabled as well.

**Comment by GULPF** [comment]
> Btw, we are already not spec compiant. To enable spec compliance the parsing of int64 needs to be disabled as well.

I don't think so actually. This is what the spec says:

>    This specification allows implementations to set limits on the range
   and precision of numbers accepted.  Since software that implements
   IEEE 754-2008 binary64 (double precision) numbers [IEEE754] is
   generally available and widely used, good interoperability can be
   achieved by implementations that expect no more precision or range
   than these provide, in the sense that implementations will
   approximate JSON numbers within the expected precision.  A JSON
   number such as 1E400 or 3.141592653589793238462643383279 may indicate
   potential interoperability problems, since it suggests that the
   software that created it expects receiving software to have greater
   capabilities for numeric magnitude and precision than is widely
   available.

The spec doesn't require that JSON numbers are treated as floats, it only suggests that using higher precision can cause bad interoperability.

**Comment by recloser** [comment]
Agreed, converting to `null` is pretty useless. I shouldn't have suggested that. On the other hand the alternative solution - storing the values as strings - seems pretty good to me. It's valid JSON so it won't cause any interop woes, and the `getFloat` proc could be modified so that it'd recognize the strings "NaN", "Infinity" and "-Infinity" and return a float in an appropriate state. This should handle the issue you described smoothly. And, unlike the solution you're proposing here, it would actually make it possible to send an inf or nan float from Nim to NimJS, which relies on the browsers builtin JSON parser.

**Comment by krux02** [comment]
I like what python does here:

``https://docs.python.org/2/library/json.html``

> If allow_nan is false (default: True), then it will be a ValueError to serialize out of range float values (nan, inf, -inf) in strict compliance of the JSON specification. If allow_nan is true, their JavaScript equivalents (NaN, Infinity, -Infinity) will be used.

**Pull request closed without merging**